### PR TITLE
Better venv handling

### PR DESCRIPTION
**Pull request by konstin**
Inspired by PEP 704 and copied from the implementation in maturin

### CODE REVIEW

**Comment by konstin** [comment]
Current dependencies on/for this PR:
* main
  * **PR #43** <a href="https://app.graphite.dev/github/pr/konstin/poc-monotrail/43" target="_blank">Graphite</a>  👈
    * **PR #44** <a href="https://app.graphite.dev/github/pr/konstin/poc-monotrail/44" target="_blank">Graphite</a> 

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/konstin/poc-monotrail/43?utm_source=stack-comment).

File: src/cli.rs [hunk 1]
```diff
@@ -421,3 +406,39 @@ pub fn run_cli(cli: Cli, venv: Option<&Path>) -> anyhow::Result<Option<i32>> {
         }
     }
 }
+
+/// Finds a) an activated venv (`VIRTUAL_ENV`) b) `.venv` in any parent folder c) tells the user
+/// about venvs
+///
+/// The optional argument allows overriding the venv in test
+pub fn find_venv(venv: Option<&Path>) -> anyhow::Result<PathBuf> {
+    let dot_venv = PathBuf::from(".venv");
```
Comment (charliermarsh) [inline src/cli.rs:415]: Can this instead be allocated on line 424?

File: src/cli.rs [hunk 2]
```diff
@@ -421,3 +406,39 @@ pub fn run_cli(cli: Cli, venv: Option<&Path>) -> anyhow::Result<Option<i32>> {
...
+        // we should check `sys.prefix == sys.base_prefix`, but we're not a python program nor
+        // running a python interpreter, in fact we want to create the environment
+        PathBuf::from(virtual_env)
+    } else if dot_venv.join("pyvenv.cfg").is_file() {
+        dot_venv
+    } else {
+        let activation_command = if cfg!(target_family = "unix") {
+            "and activate it using `source .venv/bin/activate`".to_string()
+        } else if cfg!(target_family = "windows") {
+            r#"and activate it using `.venv\Scripts\Activate.ps1`"#.to_string()
```
Comment (charliermarsh) [inline src/cli.rs:429]: Could you use `Cow` to avoid allocating here? Since only one of the branches requires an allocation.

File: src/cli.rs [hunk 3]
```diff
@@ -421,3 +406,39 @@ pub fn run_cli(cli: Cli, venv: Option<&Path>) -> anyhow::Result<Option<i32>> {
...
+        } else {
+            format!(
+                ". Please consult the documentation for {} on how to activate virtualenvs. ",
+                target_lexicon::HOST
+            )
+        };
+        bail!(
+            "Couldn't find an activated virtualenv not a .venv found in any parent directory. \
+                    You can create a virtualenv with `python -m venv .venv` {}. \
```
Comment (charliermarsh) [inline src/cli.rs:438]: Nit: I think the fact that there's a space between the .venv and `{}` means there'd be a space between .venv and the period in the `format!` case above.

**Review by charliermarsh — APPROVED** [review]

File: src/cli.rs (hunk 2 above)
Comment (konstin) [reply src/cli.rs:429]: yeah but it's the `exit(1)` path anyway

File: src/cli.rs (hunk 3 above)
Comment (konstin) [reply src/cli.rs:438]: good catch

**Pull request merged**